datasource/opennms-kafka/src/test/java/org/opennms/alec/datasource/opennms/KafkaRecoveryTest.java

```diff
@@ -0,0 +1,128 @@
+import static org.hamcrest.Matchers.hasSize;
+
+import java.lang.reflect.Method;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.streams.processor.internals.StreamThread;
+import org.junit.Test;
+import org.opennms.alec.datasource.api.Alarm;
+import org.opennms.alec.datasource.api.AlarmHandler;
+import org.opennms.alec.datasource.opennms.proto.OpennmsModelProtos;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Makes sure that we can recover after the connection to Kafka is lost. */
+public class KafkaRecoveryTest extends OpennmsDatasourceIT implements AlarmHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaRecoveryTest.class);
+
+    private final Map<String, Alarm> alarmsCreatedOrUpdatedById = new LinkedHashMap<>();
+    private final Map<String, Alarm> alarmsClearedById = new LinkedHashMap<>();
+
+    @Test
+    public void canHandleAlarms() throws Exception {
+        datasource.init();
+
+        // No alarms initially
+        assertThat(datasource.getAlarms(), hasSize(0));
+
+        // Send one alarm
+        OpennmsModelProtos.Alarm alarm = OpennmsModelProtos.Alarm.newBuilder()
+                .setReductionKey("nodeDown::1")
+                .setLastEventTime(1)
+                .setSeverity(OpennmsModelProtos.Severity.CRITICAL)
+                .build();
+        producer.send(new ProducerRecord<>(datasource.getAlarmTopic(), alarm.getReductionKey(), alarm.toByteArray()));
+
+        // Expect a single alarm to be present in the datasource
+        await().atMost(10, TimeUnit.SECONDS).until(() -> datasource.getAlarms(), hasSize(1));
+        // Now register a handler
+        assertThat(datasource.getAlarmsAndRegisterHandler(this), hasSize(1));
+        // We should not have received any callbacks yet
+        assertThat(alarmsCreatedOrUpdatedById.entrySet(), hasSize(0));
+        assertThat(alarmsClearedById.entrySet(), hasSize(0));
+
+        // Simulate a problem with KafkaStreams in OpennmsDatsource:
+        // Stop all StreamsThreads and wait for the KafkaStreamMonitor to restart the OpennmsDatasource
+        LOG.info("Trigger KafkaStream problem.");
+        // "Kill" all threads
+        for (Thread t : Thread.getAllStackTraces().keySet()) {
+            if (t instanceof StreamThread) {
+                LOG.info("Stopping Thread: " + t.getName());
+                StreamThread thread = (StreamThread)t;
+                thread.shutdown();
+                Method method = thread.getClass().getDeclaredMethod("setState", StreamThread.State.class);
+                method.setAccessible(true);
+                method.invoke(thread, StreamThread.State.DEAD); // make sure we are really dead
+            }
+        }
+
+        // lets wait a bit to allow OpennmsDatsource initialize
+        Thread.sleep(2000);
```

Comment:
Can we check that the datasource was actually re-initialized?

Reply:
:+1: done